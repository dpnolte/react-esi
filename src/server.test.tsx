import express from "express";
import React from "react";
import request from "supertest";

process.env.REACT_ESI_SECRET = "dummy";
process.env.REACT_ESI_PATH = "/_custom";
import { createIncludeElement, path, serveFragment } from "./server";

test("path", () => {
  expect(path).toBe("/_custom");
});

test("createIncludeElement", () => {
  const elem = createIncludeElement(
    "fragmentID",
    { name: "Kévin" },
    { attrs: { alt: `"'<&Alt>'"` } }
  );
  expect(elem).toMatchSnapshot();
});

const fragmentURL =
  "/_custom?fragment=fragmentID&props=%7B%22name%22%3A%22K%C3%A9vin%22%7D&sign=f7ddf06659aadbcba0cdad4c927ac5bf38167d714e1a15cad13115e7e9d21a9d";
test("serveFragment", async () => {
  const app = express();
  const resolver = (
    fragmentID: string,
    props: object,
    req: express.Request,
    res: express.Response
  ) => {
    expect(fragmentID).toBe("fragmentID");
    expect(props).toMatchObject({ name: "Kévin" });
    expect(req.header("user-agent")).toBe("test");
    expect(res).toBeDefined();

    return (p: { name: string }) => <div>Hello {p.name}</div>;
  };

  app.get(path, (req: express.Request, res: express.Response) =>
    serveFragment(req, res, resolver)
  );

  const response: any = await request(app)
    .get(fragmentURL)
    .set("user-agent", "test")
    .expect(200);
  expect(response.text).toMatchSnapshot();
});

test("serveFragment complex", async () => {
  const app = express();
  const resolver = (
    fragmentID: string,
    props: object,
    req: express.Request,
    res: express.Response
  ) => {
    expect(fragmentID).toBe("fragmentID");
    expect(props).toMatchObject({ name: "Kévin" });
    expect(req.header("user-agent")).toBe("test");
    expect(res).toBeDefined();

    return (p: { name: string }) => {
      return (
        <div className="css-a5qmhs e1e5uckb0">
          <div className="css-910pwr e1e5uckb1">
            <img
              src="/static/media/yoda.e216d82d.svg"
              className="css-r62z6h e1e5uckb2"
            />
            <h2>Page 3</h2>
          </div>
          <p className="css-iqjf8m e1e5uckb3">
            To get started, edit
            <code>src/apps/Page3/App.tsx</code>
            and save to reload.
          </p>
          <ul className="css-o9b79t e1e5uckb4">
            <li className="css-bkrwwc e1e5uckb5">
              <a href="/">Go back home</a>
            </li>
          </ul>
        </div>
      );
    };
  };

  app.get(path, (req: express.Request, res: express.Response) =>
    serveFragment(req, res, resolver)
  );

  const response: any = await request(app)
    .get(fragmentURL)
    .set("user-agent", "test")
    .expect(200);
  expect(response.text).toMatchSnapshot();
});

test("initial props", async () => {
  const app = express();
  const resolver = (
    fragmentID: string,
    props: object,
    req: express.Request,
    res: express.Response
  ) => {
    expect(fragmentID).toBe("fragmentID");
    expect(props).toMatchObject({ name: "Kévin" });
    expect(req.header("user-agent")).toBe("test");
    expect(res).toBeDefined();

    interface IPropsType {
      name: string;
    }
    const Component = (p: IPropsType) => <div>Hello {p.name}</div>;
    Component.getInitialProps = async () => {
      return { name: "Anne" };
    };

    return Component;
  };

  app.get(path, (req: express.Request, res: express.Response) =>
    serveFragment(req, res, resolver)
  );

  const response: any = await request(app)
    .get(fragmentURL)
    .set("user-agent", "test")
    .expect(200);
  expect(response.text).toMatchSnapshot();
});

test("invalid signature", async () => {
  const app = express();
  const resolver = () => () => <div />;

  app.get(path, (req: express.Request, res: express.Response) =>
    serveFragment(req, res, resolver)
  );

  const response: any = await request(app)
    .get(
      "/_custom?fragment=fragmentID&props=%7B%22foo%22%3A%22bar%22%7D&sign=invalid"
    )
    .expect(400);
  expect(response.text).toMatchSnapshot();
});
