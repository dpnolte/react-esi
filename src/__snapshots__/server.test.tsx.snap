// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createIncludeElement 1`] = `"<esi:include alt=\\"&quot;&apos;&lt;&amp;Alt&gt;&apos;&quot;\\" src=\\"/_custom?fragment=fragmentID&amp;props=%7B%22name%22%3A%22K%C3%A9vin%22%7D&amp;sign=f7ddf06659aadbcba0cdad4c927ac5bf38167d714e1a15cad13115e7e9d21a9d\\" />"`;

exports[`initial props 1`] = `"<script>window.__REACT_ESI__ = window.__REACT_ESI__ || {}; window.__REACT_ESI__['fragmentID'] = {\\"name\\":\\"Anne\\"};document.currentScript.remove();</script><div>Hello <!-- -->Anne</div>"`;

exports[`invalid signature 1`] = `"Bad signature"`;

exports[`serveFragment 1`] = `"<script>window.__REACT_ESI__ = window.__REACT_ESI__ || {}; window.__REACT_ESI__['fragmentID'] = {\\"name\\":\\"Kévin\\"};document.currentScript.remove();</script><div>Hello <!-- -->Kévin</div>"`;

exports[`serveFragment complex 1`] = `"<script>window.__REACT_ESI__ = window.__REACT_ESI__ || {}; window.__REACT_ESI__['fragmentID'] = {\\"name\\":\\"Kévin\\"};document.currentScript.remove();</script><div class=\\"css-a5qmhs e1e5uckb0\\"><div class=\\"css-910pwr e1e5uckb1\\"><img src=\\"/static/media/yoda.e216d82d.svg\\" class=\\"css-r62z6h e1e5uckb2\\"/><h2>Page 3</h2></div><p class=\\"css-iqjf8m e1e5uckb3\\">To get started, edit<code>src/apps/Page3/App.tsx</code>and save to reload.</p><ul class=\\"css-o9b79t e1e5uckb4\\"><li class=\\"css-bkrwwc e1e5uckb5\\"><a href=\\"/\\">Go back home</a></li></ul></div>"`;
